lib/utils/sudo.ts

```diff
@@ -14,34 +14,78 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-import { spawn, StdioOptions } from 'child_process';
 
 import * as Bluebird from 'bluebird';
-import * as rindle from 'rindle';
+import { ChildProcess, spawn, SpawnOptions } from 'child_process';
+import { stripIndent } from 'common-tags';
+import * as _ from 'lodash';
 
+/**
+ * Execute a child process with admin / superuser privileges, prompting
+ * the user for elevation as needed.
+ * @param command Path to an executable, plus any arguments.
+ * These will be shell-escaped by this function.
+ * @param stderr Optional stream to which stderr should be piped
+ */
 export async function executeWithPrivileges(
 	command: string[],
 	stderr?: NodeJS.WritableStream,
-): Promise<string> {
-	const stdio: StdioOptions = [
-		'inherit',
-		'inherit',
-		stderr ? 'pipe' : 'inherit',
-	];
-	const opts = {
-		env: process.env,
-		stdio,
-	};
+): Promise<void> {
+	// whether the CLI is already running with admin / super user privileges
+	const isElevated = await (await import('is-elevated'))();
+	const { shellEscape } = await import('./helpers');
+	const escapedArgs = shellEscape(command);
 
-	const args = process.argv
-		.slice(0, 2)
-		.concat(['internal', 'sudo', command.join(' ')]);
+	await new Bluebird((resolve, reject) => {
+		const opts: SpawnOptions = {
+			env: process.env,
+			shell: true,
+			stdio: ['inherit', 'inherit', stderr ? 'pipe' : 'inherit'],
+		};
+		let ps: ChildProcess;
+		let errMsgCmd: string;
 
-	const ps = spawn(args[0], args.slice(1), opts);
+		if (isElevated) {
+			// already running with privileges: simply spawn the command
+			errMsgCmd = escapedArgs.join(' ');
+			ps = spawn(escapedArgs[0], escapedArgs.slice(1), opts);
+		} else {
+			// running as ordinary user: elevate privileges
+			if (process.platform === 'win32') {
+				resolve(windosuExec(escapedArgs, stderr));
+				return;
+			}
+			errMsgCmd = ['sudo', ...escapedArgs].join(' ');
+			ps = spawn('sudo', escapedArgs, opts);
+		}
+		ps.on('error', reject);
+		ps.on('exit', codeOrSignal => {
+			if (codeOrSignal !== 0) {
+				reject(
+					new Error(
+						`Child process exited with error code "${codeOrSignal}" for command:\n${errMsgCmd}`,
+					),
+				);
+			} else {
+				resolve();
+			}
+		});
+		if (stderr) {
+			ps.stderr.pipe(stderr);
+		}
+	});
+}
 
+async function windosuExec(
+	escapedArgs: string[],
+	stderr?: NodeJS.WritableStream,
+): Promise<void> {
 	if (stderr) {
-		ps.stderr.pipe(stderr);
+		const msg = stripIndent`
+			Error: unable to elevate privileges. Please run the command prompt as an Administrator:
+			https://www.howtogeek.com/194041/how-to-open-the-command-prompt-as-administrator-in-windows-8.1/
+		`;
+		throw new Error(msg);
```

Comment:
* `windosu` can't capture stderr / stdout, hence the error for CLI commands that require stderr capture, like the interactive forms of `balena join`, `balena leave`. The user can get around it by re-executing the shell as with admin privileges.
* I pinged Juan, Lorenzon and Alexis about it: https://www.flowdock.com/app/rulemotion/resin-frontend/threads/4i6tUS12v6y-AHAMu9qT9VXSop9
* Alexis recommended `sudo-prompt` as an alternative to `windosu`. I tested it, and found that while `windosu` doesn't capture enough, `sudo-prompt` captures too much: captures and buffers (`Buffer` / `string` rather than `stream`) stdout and stderr. This is a problem because the elevated privilege child process is expected to interact with the user via stdin/stdout (interactive text menus) before producing a capturable output in stderr. (The whole design is questionable, but it's how it is.)
* I actually had a go at patching both/either `windosu` and `sudo-prompt`, perhaps against best advice (I blame Page, :-) because of his defence of user-friendliness of the on-demand privilege elevation on Windows, versus the inconvenient of re-starting the shell as admin...) Sadly it proved trickier than I had hoped. The `sudo-prompt` package is more promising in terms of patching (for future attempts), but both `windosu` and `sudo-prompt` spawn child processes that would have to align / coordinate with the CLI's objective of capturing stderr only, leaving stdin and stdout alone. It sounds like a simple objective, but it proved a hair-pulling sort of debugging and I eventually gave up. As discussed in the previous CLI check-in, this had never worked before on Windows (stderr capture), so even the advice of re-executing the shell as admin is an improvement, and we have higher-priority work to focus on.